lua/barbar/render.lua

```diff
@@ -139,6 +139,31 @@ local function hl_tabline(group)
   return '%#' .. group .. '#'
 end
 
+-- Digits for optional styling of buffer_number and buffer_index.
+local SUPERSCRIPT_DIGITS = { '⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹' }
+local SUBSCRIPT_DIGITS = { '₀', '₁', '₂', '₃', '₄', '₅', '₆', '₇', '₈', '₉' }
+
+---@param num number
+---@param style barbar.config.options.icons.numbers.value
+local function style_number(num, style)
+  local digits
+  local num_as_str = tostring(num)
+  if style == 'subscript' then
+    digits = SUBSCRIPT_DIGITS
+  elseif style == 'superscript' then
+    digits = SUPERSCRIPT_DIGITS
+  else
+    return num_as_str
+  end
```

Comment:
You know what, pass those digits as an argument to `style_number` and we can keep the string constants. E.g. `style_number(num, style, digits)`. The digits would be either a table or nil. The string comparison happens once per render, I'm cool with that.

Reply:
Good call, that can be resolved outside the loop.

Reply:
Wouldn't this give us two of those?

```lua
buffer_index.text = style_number(
  i,
  icons_option.buffer_index,
  icons_option.buffer_index == 'superscript' and SUPERSCRIPT_DIGITS
    or icons_option.buffer_index == 'subscript' and SUBSCRIPT_DIGITS
    or nil) .. ' '
```

The partial reverting the full encapsulation I did in the second commit might be more elegant. Only calling style_number() when superscript or subscript is specified but not passing digits as arg.
```lua
buffer_number.text = ((icons_option.buffer_number == "superscript" or icons_option.buffer_number == "subscript")
        and style_number(bufnr, icons_option.buffer_number)
        or tostring(bufnr)) .. ' '
```

What you think?

Reply:
You'd assign digits outside of the loop.

```lua
local digits = nil
if opt == 'super' then digits = SUPER end
if opt == 'sub' then digits = SUB end

for ...
  text = style_number(..., digits)
end
```

Reply:
I see, didn't realized the outer loop. Thought the loop that needed to be solved was solved with iron-e suggestions. Thanks!

Reply:
Problem is, to assign the digits, we need access to 

```lua
local icons_option = state.icons(bufnr, activity)
```

Which is inside the loop.

Reply:
Ah, right. Because the icons of a buffer depends on the buffer number / activity of the buffer.

Reply:
I think our options are too complex if we need to be in a loop interation to know if there are script digits. @Iron-E wdyt? Could we simplify things a bit here?

Reply:
Technically we always needed to be in a loop to resolve what the current options were. We've had e.g. `icon_inactive_separator` since <1.0.0, and you need to do `Buffer.get_activity` in the `generate_tabline` loop to determine whether to use that option.

We've had a lot of requests to change the appearance for certain types of buffers to look one way different from the norm (e.g. #217, #97 which I disable for current buffers, #379, and most recently #408) and I myself [use the feature as well](https://gitlab.com/Iron_E/dotfiles/-/blob/master/.config/nvim/lua/init/lazy.lua#L306-L321). 

Though interestingly, #386 actually increased performance quite a bit even though we had this increased complexity. 

I have a couple ideas that could increase performance, but it'd have to be in a followup PR because it touches a lot of things

Reply:
And there is no way to have that option outside the buffer-specific options? I really doubt someone would want superscript somewhere and subscript elsewhere. If not that's fine for now.

Reply:
Oh, if that's what you mean, we could totally do something like `buffer_index_style` and `buffer_number_style`.

Reply:
@romgrk see 9fc4ac50e02866637b1599a95b21de648a02ce44. I implemented a 2× performance increase

Reply:
I thought not jumping into ongoing changes and waiting is the right move on my end. Would it be beneficial to make further changes like in @Iron-E's linked comimt or should I wait until the options are out of the for loop on your side first?

Reply:
No matter what, certain options have to be resolved in the loop. That commit just increased the perf of doing that.

If @romgrk is okay with it, you can implement the with `style_numbers` resolving digits internally and do a benchmark to see what the impact is.

You can use [this code](https://gitlab.com/Iron_E/dotfiles/-/blob/master/.config/nvim/lua/init/config.lua#L173) to do that

Reply:
Thanks for the snip! On my box I just had little impact during normal usage while running style number in a loops of 1000.
Highest value after a thousand loops `0.0036389999999997`. Lowest value `0.00029000000000146`

Reply:
Thanks for doing that. We can let romgrk decide from here 🙂

Reply:
Let's pick whatever option is more simple, doesn't seem to have a significant performance impact.

Reply:
I think the state now that includes the suggestion after the initial PR is fairly concise and readable 🙂:+1: